### PR TITLE
Feature/product actions

### PR DESCRIPTION
**Pull request by marustas**

### CODE REVIEW

File: src/app/pages/fridge/components/add-product/add-product.component.ts [hunk 1]
```diff
@@ -25,14 +28,57 @@ export class AddProductComponent {
...
+    quantity = unit === 'N/A' || 'undefined' ? splitQuantity : quantity;
+
+    if (!newProductExpiration) {
+      const currentDate = new Date();
+      currentDate.setDate(currentDate.getDate() + 12);
+      const day = currentDate.getDate();
+      const month = currentDate.getMonth() + 1;
+      const year = currentDate.getFullYear();
+      newProductExpiration = `${day}/${month}/${year}`;
+    }
```
Comment (Howl404) [inline src/app/pages/fridge/components/add-product/add-product.component.ts]: Make util function

File: src/app/pages/fridge/components/add-product/add-product.component.ts [hunk 2]
```diff
@@ -25,14 +28,57 @@ export class AddProductComponent {
...
+
+    const newProduct: Ingredient = {
+      id: 0,
+      quantity: quantity,
```
Comment (Howl404) [inline src/app/pages/fridge/components/add-product/add-product.component.ts]:
```suggestion
      quantity,
```

File: src/app/pages/fridge/components/add-product/add-product.component.ts [hunk 3]
```diff
@@ -25,14 +28,57 @@ export class AddProductComponent {
...
+      name: newProductName,
+      expiresAt: newProductExpiration,
+    };
+
+    if (this.productID) {
+      newProduct.id = +this.productID;
+    } else {
+      newProduct.id = Math.floor(Math.random() * 1000) + this.productService.currentLength;
```
Comment (Howl404) [inline src/app/pages/fridge/components/add-product/add-product.component.ts:73]: Util function (getRandomId)

File: src/app/pages/fridge/components/products-cards/products-cards.component.html [hunk 4]
```diff
@@ -1,6 +1,17 @@
-<div class="cards-container">
-  <app-product-card
-    *ngFor="let product of products$ | async"
-    [product]="product"
-  ></app-product-card>
-</div>
+<ng-container *ngIf="products$ | async as products; else spinnerTemplate">
+  <div class="cards-container" *ngIf="products.length > 0; else noDataTemplate">
```
Comment (Howl404) [inline src/app/pages/fridge/components/products-cards/products-cards.component.html]:
```suggestion
  <div class="cards-container" *ngIf="products.length; else noDataTemplate">
```

File: src/app/pages/fridge/components/products-table/products-table.component.html [hunk 5]
```diff
@@ -1,10 +1,42 @@
...
-  </tr>
-  <tr *ngFor="let product of products$ | async">
-    <td>{{ product.name }}</td>
-    <td>{{ product.quantity }}</td>
-    <td>{{ product.expiresAt | stringToDate | date: 'd MMM, yyyy' }}</td>
-  </tr>
-</table>
+<ng-container *ngIf="products$ | async as products; else spinnerTemplate">
+  <table
+    *ngIf="products.length > 0; else noDataTemplate"
```
Comment (Howl404) [inline src/app/pages/fridge/components/products-table/products-table.component.html]:
```suggestion
    *ngIf="products.length; else noDataTemplate"
```

File: src/app/pages/fridge/components/products-table/products-table.component.html [hunk 6]
```diff
@@ -1,10 +1,42 @@
...
+    <tr
+      *ngFor="let product of products"
+      (mouseenter)="onToggleHover(product.id, true)"
+      (mouseleave)="onToggleHover(product.id, false)"
+    >
+      <td>{{ product.name }}</td>
+      <td>{{ product.quantity }}</td>
+      <td>{{ product.expiresAt | stringToDate | date: 'd MMM, yyyy' }}</td>
+      <div *ngIf="product.id === hoverIndex" class="action-icons">
+        <button class="edit-btn" (click)="onEditProduct()">
```
Comment (Howl404) [inline src/app/pages/fridge/components/products-table/products-table.component.html:14]: type="button"

File: src/app/pages/fridge/components/products-table/products-table.component.html [hunk 7]
```diff
@@ -1,10 +1,42 @@
...
+    *ngIf="products.length > 0; else noDataTemplate"
+    class="products-table"
+  >
+    <tr>
+      <th *ngFor="let title of titles">{{ title | uppercase }}</th>
+    </tr>
+    <tr
+      *ngFor="let product of products"
+      (mouseenter)="onToggleHover(product.id, true)"
+      (mouseleave)="onToggleHover(product.id, false)"
```
Comment (Howl404) [inline src/app/pages/fridge/components/products-table/products-table.component.html]: Delete, nothing happens

File: src/app/pages/fridge/components/products-table/products-table.component.ts [hunk 8]
```diff
@@ -11,9 +14,40 @@ import { SearchService } from '../../services/search.service';
...
+    this.productsService.deleteProduct(this.hoverIndex);
+    this.products$ = this.productsService.products$;
+    this.hideModal();
+  }
+
+  showModal(): void {
+    this.isModalVisible = true;
+  }
+
+  hideModal(): void {
```
Comment (Howl404) [inline src/app/pages/fridge/components/products-table/products-table.component.ts:46]: private

File: src/app/pages/fridge/components/products-table/products-table.component.ts [hunk 9]
```diff
@@ -11,9 +14,40 @@ import { SearchService } from '../../services/search.service';
...
-  constructor(searchService: SearchService) {
-    this.products$ = searchService.searchQuery$.pipe(switchMap((query) => searchService.searchProducts(query)));
+  constructor(
+    searchService: SearchService,
+    private productsService: ProductsService,
+    loaderService: LoaderService,
+    private router: Router
+  ) {
+    this.products$ = searchService.searchQuery$.pipe(switchMap((query) => searchService.searchProducts(query))).pipe();
+    loaderService.isLoading$.subscribe((isLoading) => (this.isLoading = isLoading));
```
Comment (Howl404) [inline src/app/pages/fridge/components/products-table/products-table.component.ts]: Unsubscribe

File: src/app/pages/fridge/components/add-product/add-product.component.ts [hunk 10]
```diff
@@ -25,14 +28,59 @@ export class AddProductComponent {
     });
   }
 
+  ngOnInit(): void {
+    this.route.paramMap.subscribe((params) => {
```
Comment (alesia-bulka) [inline src/app/pages/fridge/components/add-product/add-product.component.ts:32]: UNSUBSCRIBE

**Pull request merged**